File: my/hybrid_astar/src/obstacle_heuristic.cpp
#include <algorithm>
#include <cmath>

#include "hybrid_astar/obstacle_heuristic.hpp"
#include "hybrid_astar/costmap_2d.hpp"
#include "hybrid_astar/esdf_holder.hpp"

namespace hybrid_astar
{

void ObstacleHeuristic::resetObstacleHeuristic(
  Costmap2D * costmap,
  const float & start_x, const float & start_y,
  const float & goal_x, const float & goal_y,
  const bool downsample_obstacle_heuristic)
{
  this->costmap = costmap;

  unsigned int size = 0u;
  if (downsample_obstacle_heuristic) {
    cached_size_x_ = ceil(static_cast<float>(costmap->getSizeInCellsX()) / 2.0f);
    cached_size_y_ = ceil(static_cast<float>(costmap->getSizeInCellsY()) / 2.0f);
  } else {
    cached_size_x_ = costmap->getSizeInCellsX();
    cached_size_y_ = costmap->getSizeInCellsY();
  }
  size = cached_size_x_ * cached_size_y_;

  if (obstacle_heuristic_lookup_table_.size() == size) {
    std::fill(
      obstacle_heuristic_lookup_table_.begin(),
      obstacle_heuristic_lookup_table_.end(), 0.0f);
  } else {
    obstacle_heuristic_lookup_table_.resize(size, 0.0f);
    std::fill(
      obstacle_heuristic_lookup_table_.begin(),
      obstacle_heuristic_lookup_table_.end(), 0.0f);
  }

  obstacle_heuristic_queue_.clear();
  obstacle_heuristic_queue_.reserve(size);

  const float goal_x_floor = std::max(0.0f, std::floor(goal_x));
  const float goal_y_floor = std::max(0.0f, std::floor(goal_y));
  unsigned int gx = static_cast<unsigned int>(goal_x_floor);
  unsigned int gy = static_cast<unsigned int>(goal_y_floor);
  if (gx >= cached_size_x_) {gx = cached_size_x_ - 1u;}
  if (gy >= cached_size_y_) {gy = cached_size_y_ - 1u;}
  unsigned int goal_index;
  if (downsample_obstacle_heuristic) {
    goal_index = (gy / 2u) * cached_size_x_ + (gx / 2u);
  } else {
    goal_index = gy * cached_size_x_ + gx;
  }

  const float inv = downsample_obstacle_heuristic ? 2.0f : 1.0f;
  const float start_x_floor = std::floor(start_x / inv);
  const float start_y_floor = std::floor(start_y / inv);
  obstacle_heuristic_queue_.emplace_back(
    distanceHeuristic2D(
      goal_index, cached_size_x_,
      static_cast<unsigned int>(start_x_floor),
      static_cast<unsigned int>(start_y_floor)),
    goal_index);

  obstacle_heuristic_lookup_table_[goal_index] = -0.00001f;
}

float ObstacleHeuristic::getObstacleHeuristic(
  const Coordinates & node_coords,
  const float & cost_penalty,
  const bool use_quadratic_cost_penalty,
  const bool downsample_obstacle_heuristic)
{
  const unsigned int size_x = cached_size_x_;
  const unsigned int size_y = cached_size_y_;

  float start_x_f, start_y_f;
  if (downsample_obstacle_heuristic) {
    start_y_f = std::floor(node_coords.y / 2.0f);
    start_x_f = std::floor(node_coords.x / 2.0f);
  } else {
    start_y_f = std::floor(node_coords.y);
    start_x_f = std::floor(node_coords.x);
  }
  if (start_x_f < 0.0f) {start_x_f = 0.0f;}
  if (start_y_f < 0.0f) {start_y_f = 0.0f;}
  const unsigned int start_x = static_cast<unsigned int>(start_x_f);
  const unsigned int start_y = static_cast<unsigned int>(start_y_f);

  const unsigned int start_index = start_y * size_x + start_x;
  const float & requested_node_cost = obstacle_heuristic_lookup_table_[start_index];
  if (requested_node_cost > 0.0f) {
    return downsample_obstacle_heuristic ? 2.0f * requested_node_cost : requested_node_cost;
  }

  for (auto & n : obstacle_heuristic_queue_) {
    n.first = -obstacle_heuristic_lookup_table_[n.second] +
      distanceHeuristic2D(n.second, size_x, start_x, start_y);
  }
  std::make_heap(
    obstacle_heuristic_queue_.begin(), obstacle_heuristic_queue_.end(),
    NodeHeuristicComparator{});

  const int size_x_int = static_cast<int>(size_x);
  const float sqrt2 = sqrtf(2.0f);
  float c_cost, cost, travel_cost, new_cost, existing_cost;
  unsigned int mx, my;
  unsigned int idx, new_idx = 0;

  const std::vector<int> neighborhood = {1, -1,
    size_x_int, -size_x_int,
    size_x_int + 1, size_x_int - 1,
    -size_x_int + 1, -size_x_int - 1};

  while (!obstacle_heuristic_queue_.empty()) {
    idx = obstacle_heuristic_queue_.front().second;
    std::pop_heap(
      obstacle_heuristic_queue_.begin(), obstacle_heuristic_queue_.end(),
      NodeHeuristicComparator{});
    obstacle_heuristic_queue_.pop_back();
    c_cost = obstacle_heuristic_lookup_table_[idx];
    if (c_cost > 0.0f) {
      continue;
    }
    c_cost = -c_cost;
    obstacle_heuristic_lookup_table_[idx] = c_cost;

    for (unsigned int i = 0; i != neighborhood.size(); i++) {
      int new_idx_int = static_cast<int>(idx) + neighborhood[i];
      if (new_idx_int < 0 || new_idx_int >= static_cast<int>(size_x * size_y)) {
        continue;
      }
      new_idx = static_cast<unsigned int>(new_idx_int);

      const int new_my = static_cast<int>(new_idx / size_x);
      const int new_mx_signed = static_cast<int>(new_idx - static_cast<unsigned int>(new_my) * size_x);
      if (new_mx_signed < 0 || static_cast<unsigned int>(new_mx_signed) >= size_x ||
        new_my < 0 || static_cast<unsigned int>(new_my) >= size_y)
      {
        continue;
      }

      {
        if (downsample_obstacle_heuristic) {
          unsigned int y_offset = (new_idx / size_x) * 2;
          unsigned int x_offset = (new_idx - ((new_idx / size_x) * size_x)) * 2;
          cost = cellCostForHeuristic(x_offset, y_offset);
          for (unsigned int k = 0; k < 2u; ++k) {
            unsigned int mxd = x_offset + k;
            if (mxd >= costmap->getSizeInCellsX()) {
              continue;
            }
            for (unsigned int j = 0; j < 2u; ++j) {
              unsigned int myd = y_offset + j;
              if (myd >= costmap->getSizeInCellsY()) {
                continue;
              }
              if (k == 0 && j == 0) {
                continue;
              }
              cost = std::min(cost, cellCostForHeuristic(mxd, myd));
            }
          }
        } else {
          cost = cellCostForHeuristic(new_idx % size_x, new_idx / size_x);
        }

        if (cost >= INSCRIBED_COST) {
          continue;
        }

        my = static_cast<unsigned int>(new_my);
        mx = static_cast<unsigned int>(new_mx_signed);

        if (size_x <= 3u || size_y <= 3u) {
          continue;
        }
        if (mx <= 3u || mx >= size_x - 3u) {
          continue;
        }
        if (my <= 3u || my >= size_y - 3u) {
          continue;
        }

        existing_cost = obstacle_heuristic_lookup_table_[new_idx];
        if (existing_cost <= 0.0f) {
          if (use_quadratic_cost_penalty) {
            travel_cost =
              (i <= 3 ? 1.0f : sqrt2) * (1.0f + (cost_penalty * cost * cost / MAX_NON_OBSTACLE_COST_SQ));
          } else {
            travel_cost =
              ((i <= 3) ? 1.0f : sqrt2) * (1.0f + (cost_penalty * cost / MAX_NON_OBSTACLE_COST));
          }

          new_cost = c_cost + travel_cost;
          if (existing_cost == 0.0f || -existing_cost > new_cost) {
            obstacle_heuristic_lookup_table_[new_idx] = -new_cost;
            obstacle_heuristic_queue_.emplace_back(
              new_cost + distanceHeuristic2D(new_idx, size_x, start_x, start_y), new_idx);
            std::push_heap(
              obstacle_heuristic_queue_.begin(), obstacle_heuristic_queue_.end(),
              NodeHeuristicComparator{});
          }
        }
      }
    }

    if (idx == start_index) {
      break;
    }
  }
  return downsample_obstacle_heuristic ? 2.0f * requested_node_cost : requested_node_cost;
}

float ObstacleHeuristic::cellCostForHeuristic(unsigned int mx, unsigned int my)
{
  // Legacy path: use the raw costmap cell cost.
  if (esdf_holder_ == nullptr || !esdf_holder_->valid()) {
    return static_cast<float>(costmap->getCost(mx, my));
  }

  // ESDF path: compute the soft penalty at the cell center using the
  // capsule footprint (or a single-circle fallback). The footprint is
  // queried without rotation (theta=0): the heuristic is a 2D distance
  // field independent of orientation. The penalty is mapped to
  // [0, MAX_NON_OBSTACLE_COST] so the existing cost_penalty weighting
  // scales correctly.
  const double resolution = costmap->getResolution();
  const double wx = costmap->getOriginX() + (static_cast<double>(mx) + 0.5) * resolution;
  const double wy = costmap->getOriginY() + (static_cast<double>(my) + 0.5) * resolution;
  double min_clearance = esdf_holder_->clearanceAtWorld(wx, wy);
  if (!cost_check_points_.empty()) {
    min_clearance = std::numeric_limits<double>::infinity();
    for (size_t offset = 0; offset + 2 < cost_check_points_.size(); offset += 3) {
      const double lx = cost_check_points_[offset + 0];
      const double ly = cost_check_points_[offset + 1];
      const double d = esdf_holder_->clearanceAtWorld(wx + lx, wy + ly);
      if (d < min_clearance) {
        min_clearance = d;
      }
    }
  }

  if (!std::isfinite(min_clearance)) {
    return static_cast<float>(MAX_NON_OBSTACLE_COST);
  }
  const double surface_distance = min_clearance - robot_radius_;
  if (surface_distance >= safe_distance_) {
    return 0.0f;
  }
  const double normalized_gap = (safe_distance_ - surface_distance) / safe_distance_;
  const double penalty = normalized_gap * normalized_gap;
  return static_cast<float>(penalty * MAX_NON_OBSTACLE_COST);
}

}  // namespace hybrid_astar

File: my/hybrid_astar_rs/src/obstacle_heuristic.rs
// Copyright 2024–2026  The hybrid_astar Contributors
// SPDX-License-Identifier: MIT

//! D*-style 2-D obstacle distance heuristic.

use std::cell::RefCell;

use crate::constants::*;
use crate::costmap::Costmap2D;
use crate::esdf::EsdfHolder;
use crate::types::*;

/// D*-style obstacle heuristic that pre-computes a 2-D distance field from
/// the goal cell.
pub struct ObstacleHeuristic {
    lookup_table: RefCell<Vec<f32>>,
    queue: RefCell<Vec<NodeHeuristicPair>>,
    costmap: Option<*const Costmap2D>,
    cached_size_x: u32,
    cached_size_y: u32,

    // Optional ESDF path
    esdf_holder: Option<*const EsdfHolder>,
    cost_check_points: Vec<f64>,
    robot_radius: f64,
    safe_distance: f64,
}

// Safety: used single-threaded.
unsafe impl Send for ObstacleHeuristic {}

impl ObstacleHeuristic {
    pub fn new() -> Self {
        Self {
            lookup_table: RefCell::new(Vec::new()),
            queue: RefCell::new(Vec::new()),
            costmap: None,
            cached_size_x: 0,
            cached_size_y: 0,
            esdf_holder: None,
            cost_check_points: Vec::new(),
            robot_radius: 0.0,
            safe_distance: 0.0,
        }
    }

    pub fn set_esdf_holder(&mut self, holder: *const EsdfHolder) {
        self.esdf_holder = Some(holder);
    }

    pub fn set_esdf_footprint_params(
        &mut self,
        cost_check_points: Vec<f64>,
        robot_radius: f64,
        safe_distance: f64,
    ) {
        self.cost_check_points = cost_check_points;
        self.robot_radius = robot_radius;
        self.safe_distance = safe_distance;
    }

    pub fn reset_obstacle_heuristic(
        &mut self,
        costmap: &Costmap2D,
        start_x: f32,
        start_y: f32,
        goal_x: f32,
        goal_y: f32,
        downsample_obstacle_heuristic: bool,
    ) {
        self.costmap = Some(costmap);

        let (size_x, size_y) = if downsample_obstacle_heuristic {
            (
                ((costmap.size_x() as f32) / 2.0).ceil() as u32,
                ((costmap.size_y() as f32) / 2.0).ceil() as u32,
            )
        } else {
            (costmap.size_x(), costmap.size_y())
        };
        self.cached_size_x = size_x;
        self.cached_size_y = size_y;
        let size = (size_x * size_y) as usize;

        let mut lt = self.lookup_table.borrow_mut();
        lt.resize(size, 0.0);
        lt.iter_mut().for_each(|v| *v = 0.0);

        let mut q = self.queue.borrow_mut();
        q.clear();
        q.reserve(size);

        let goal_x_floor = goal_x.floor().max(0.0) as u32;
        let goal_y_floor = goal_y.floor().max(0.0) as u32;
        let gx = goal_x_floor.min(size_x - 1);
        let gy = goal_y_floor.min(size_y - 1);

        let goal_index = if downsample_obstacle_heuristic {
            (gy / 2) * size_x + (gx / 2)
        } else {
            gy * size_x + gx
        };

        let inv = if downsample_obstacle_heuristic { 2.0 } else { 1.0 };
        let start_x_floor = (start_x / inv).floor();
        let start_y_floor = (start_y / inv).floor();

        q.push((
            self.distance_heuristic_2d(goal_index, size_x, start_x_floor as u32, start_y_floor as u32),
            goal_index as u64,
        ));
        lt[goal_index as usize] = -0.00001;
    }

    pub fn get_obstacle_heuristic(
        &self,
        node_coords: Coordinates,
        cost_penalty: f32,
        use_quadratic_cost_penalty: bool,
        downsample_obstacle_heuristic: bool,
    ) -> f32 {
        let size_x = self.cached_size_x;
        let size_y = self.cached_size_y;

        let (start_x_f, start_y_f) = if downsample_obstacle_heuristic {
            (node_coords.x.floor() / 2.0, node_coords.y.floor() / 2.0)
        } else {
            (node_coords.x.floor(), node_coords.y.floor())
        };
        let start_x = (start_x_f.max(0.0)) as u32;
        let start_y = (start_y_f.max(0.0)) as u32;
        let start_index = start_y * size_x + start_x;

        let lt = self.lookup_table.borrow();
        let requested_node_cost = lt[start_index as usize];
        drop(lt);

        if requested_node_cost > 0.0 {
            return if downsample_obstacle_heuristic {
                2.0 * requested_node_cost
            } else {
                requested_node_cost
            };
        }

        let size_x_int = size_x as i32;
        let sqrt2 = 2.0_f32.sqrt();
        let neighborhood: [i32; 8] = [
            1, -1, size_x_int, -size_x_int, size_x_int + 1, size_x_int - 1,
            -size_x_int + 1, -size_x_int - 1,
        ];

        let mut lt = self.lookup_table.borrow_mut();
        let mut q = self.queue.borrow_mut();

        while !q.is_empty() {
            let mut min_idx = 0;
            for i in 1..q.len() {
                if q[i].0 < q[min_idx].0 {
                    min_idx = i;
                }
            }
            let node_idx = q[min_idx].1;
            q.swap_remove(min_idx);

            let c_cost = lt[node_idx as usize];
            if c_cost > 0.0 {
                continue;
            }
            let c_cost = -c_cost;
            lt[node_idx as usize] = c_cost;

            for (i, &nbr) in neighborhood.iter().enumerate() {
                let new_idx_int = node_idx as i32 + nbr;
                if new_idx_int < 0 || new_idx_int >= (size_x * size_y) as i32 {
                    continue;
                }
                let new_idx = new_idx_int as u32;

                let new_my = new_idx / size_x;
                let new_mx = new_idx - new_my * size_x;
                if new_mx >= size_x || new_my >= size_y {
                    continue;
                }

                let cost = self.cell_cost_for_heuristic(new_mx, new_my);
                if cost >= INSCRIBED_COST {
                    continue;
                }

                if size_x <= 3 || size_y <= 3 {
                    continue;
                }
                if new_mx <= 3 || new_mx >= size_x - 3 {
                    continue;
                }
                if new_my <= 3 || new_my >= size_y - 3 {
                    continue;
                }

                let existing_cost = lt[new_idx as usize];
                if existing_cost <= 0.0 {
                    let travel_cost = if use_quadratic_cost_penalty {
                        (if i <= 3 { 1.0 } else { sqrt2 })
                            * (1.0 + cost_penalty * cost * cost / MAX_NON_OBSTACLE_COST_SQ)
                    } else {
                        (if i <= 3 { 1.0 } else { sqrt2 })
                            * (1.0 + cost_penalty * cost / MAX_NON_OBSTACLE_COST)
                    };

                    let new_cost = c_cost + travel_cost;
                    if existing_cost == 0.0 || -existing_cost > new_cost {
                        lt[new_idx as usize] = -new_cost;
                        q.push((
                            new_cost + self.distance_heuristic_2d(new_idx, size_x, start_x, start_y),
                            new_idx as u64,
                        ));
                    }
                }
            }

            if node_idx == start_index as u64 {
                break;
            }
        }

        let cost = lt[start_index as usize];
        if downsample_obstacle_heuristic {
            2.0 * cost
        } else {
            cost
        }
    }

    fn cell_cost_for_heuristic(&self, mx: u32, my: u32) -> f32 {
        if self.esdf_holder.is_none() {
            let cm = unsafe { &*self.costmap.unwrap() };
            return cm.get_cost(mx, my) as f32;
        }

        let holder = unsafe { &*self.esdf_holder.unwrap() };
        if !holder.valid() {
            let cm = unsafe { &*self.costmap.unwrap() };
            return cm.get_cost(mx, my) as f32;
        }

        let cm = unsafe { &*self.costmap.unwrap() };
        let resolution = cm.resolution();
        let wx = cm.origin_x() + (mx as f64 + 0.5) * resolution;
        let wy = cm.origin_y() + (my as f64 + 0.5) * resolution;

        let mut min_clearance = holder.clearance_at_world(wx, wy);
        if !self.cost_check_points.is_empty() {
            min_clearance = f64::INFINITY;
            let mut offset = 0;
            while offset + 2 < self.cost_check_points.len() {
                let lx = self.cost_check_points[offset];
                let ly = self.cost_check_points[offset + 1];
                let d = holder.clearance_at_world(wx + lx, wy + ly);
                if d < min_clearance {
                    min_clearance = d;
                }
                offset += 3;
            }
        }

        if !min_clearance.is_finite() {
            return MAX_NON_OBSTACLE_COST;
        }
        let surface_distance = min_clearance - self.robot_radius;
        if surface_distance >= self.safe_distance {
            return 0.0;
        }
        let normalized_gap = (self.safe_distance - surface_distance) / self.safe_distance;
        let penalty = normalized_gap * normalized_gap;
        (penalty * MAX_NON_OBSTACLE_COST as f64) as f32
    }

    #[inline]
    fn distance_heuristic_2d(&self, idx: u32, size_x: u32, target_x: u32, target_y: u32) -> f32 {
        let dx = (idx % size_x) as i32 - target_x as i32;
        let dy = (idx / size_x) as i32 - target_y as i32;
        ((dx * dx + dy * dy) as f32).sqrt()
    }
}
